x-pack/plugins/maps/public/reducers/map/map.ts

```diff
@@ -94,6 +97,26 @@ export function map(state: MapState = DEFAULT_MAP_STATE, action: Record<string,
           drawState: action.drawState,
         },
       };
+    case UPDATE_EDIT_STATE_LAYER:
+      const editState = action.layerId ? { layerId: action.layerId } : undefined;
+      return {
+        ...state,
+        mapState: {
+          ...state.mapState,
+          editState,
+        },
+      };
+    case UPDATE_EDIT_STATE_SHAPE:
+      return {
+        ...state,
+        mapState: {
+          ...state.mapState,
+          editState: {
+            ...state.mapState.editState,
```

Comment:
This spread could cause an error since "state.mapState.editState" could be undefined.

How about just having a single reducer `SET_EDIT_STATE`.

so the reducer would just be
```
case SET_EDIT_STATE:
      return {
        ...state,
        mapState: {
          ...state.mapState,
          editState: action.editState,
        },
      };
```

setDrawMode would be
```
export function setDrawMode(drawMode: DRAW_MODE) {
  return (dispatch: ThunkDispatch<MapStoreState, void, AnyAction>) => {
    if (drawMode === DRAW_MODE.NONE) {
      dispatch({
        type: SET_EDIT_STATE,
        editState: undefined,
      });
    }
    dispatch({
      type: SET_DRAW_MODE,
      drawMode,
    });
  };
}
```

updateEditLayer action would be
```
export function updateEditLayer(layerId: string | null) {
  return (dispatch: Dispatch) => {
    if (layerId !== null) {
      dispatch({ type: SET_OPEN_TOOLTIPS, openTooltips: [] });
    }
    dispatch({
      type: SET_DRAW_MODE,
      drawMode: DRAW_MODE.NONE,
    });
    dispatch({
      type: SET_EDIT_STATE,
      editState: layerId ? { layerId } : undefined,
    });
  };
}
```

and updateEditShape
```
export function updateEditShape(shapeToDraw: DRAW_SHAPE | null) {
  return (dispatch: Dispatch, getState: () => MapStoreState) => {
    const editState = getEditState(getState());
    if (!editState) {
      return;
    }
    dispatch({
      type: SET_EDIT_STATE,
      editState: {
        ...editState,
        drawShape: shapeToDraw,
      },
    });
  };
}
```

Reply:
Thanks for the breakdown. Done!